p2p/netaddress.go

```diff
@@ -28,25 +31,45 @@ type NetAddress struct {
 // using 0.0.0.0:0. When normal run, other net.Addr (except TCP) will
 // panic.
 // TODO: socks proxies?
-func NewNetAddress(addr net.Addr) *NetAddress {
+func NewNetAddress(id ID, addr net.Addr) *NetAddress {
 	tcpAddr, ok := addr.(*net.TCPAddr)
 	if !ok {
 		if flag.Lookup("test.v") == nil { // normal run
 			cmn.PanicSanity(cmn.Fmt("Only TCPAddrs are supported. Got: %v", addr))
 		} else { // in testing
-			return NewNetAddressIPPort(net.IP("0.0.0.0"), 0)
+			netAddr := NewNetAddressIPPort(net.IP("0.0.0.0"), 0)
+			netAddr.ID = id
+			return netAddr
 		}
 	}
 	ip := tcpAddr.IP
 	port := uint16(tcpAddr.Port)
-	return NewNetAddressIPPort(ip, port)
+	netAddr := NewNetAddressIPPort(ip, port)
+	netAddr.ID = id
+	return netAddr
 }
 
 // NewNetAddressString returns a new NetAddress using the provided
 // address in the form of "IP:Port". Also resolves the host if host
 // is not an IP.
 func NewNetAddressString(addr string) (*NetAddress, error) {
-	host, portStr, err := net.SplitHostPort(removeProtocolIfDefined(addr))
+	addr = removeProtocolIfDefined(addr)
+
+	var id ID
+	spl := strings.Split(addr, "@")
+	if len(spl) == 2 {
+		idStr := spl[0]
+		idBytes, err := hex.DecodeString(idStr)
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("Address (%s) contains invalid ID", addr))
+		}
+		if len(idBytes) != 20 {
```

Comment:
maybe create a `const peerIDSize = 20`?

Reply:
good catch. So far the addresses we use in go-crypto are 20 bytes but nothing really guarantees that for future addresses. Maybe we will make that convention but for now, I'll simply remove this check